File: src/main/javabak/tile_entities/capabilities/MergedIoItemCapability.java
package redcrafter07.processed.block.tile_entities.capabilities;

import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.items.IItemHandler;
import net.neoforged.neoforge.items.IItemHandlerModifiable;
import net.neoforged.neoforge.items.wrapper.EmptyItemHandler;
import redcrafter07.processed.block.machine_abstractions.IoState;
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine;

public class MergedIoItemCapability implements IItemHandlerModifiable {

    private final ProcessedMachine.CapabilityHandlers handlers;

    public MergedIoItemCapability(ProcessedMachine.CapabilityHandlers handlers) {
        this.handlers = handlers;
    }

    private IItemHandler get(IoState state) {
        final var handler = handlers.getItemHandlerForState(state);
        if (handler == null) return EmptyItemHandler.INSTANCE;
        return handler;
    }

    @Override
    public int getSlots() {
        return get(IoState.Output).getSlots() + get(IoState.Input).getSlots();
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        final var handler = handlers.getItemHandlerForState(IoState.Output);
        if (handler == null) return ItemStack.EMPTY;
        if (slot >= handler.getSlots() || slot < 0) return ItemStack.EMPTY;
        return handler.getStackInSlot(slot);
    }

    @Override
    public void setStackInSlot(int slot, ItemStack stack) {
        var realSlot = slot - get(IoState.Output).getSlots();

        if (get(IoState.Input) instanceof IItemHandlerModifiable handler) {
            if (realSlot < 0 || realSlot >= handler.getSlots()) return;
            handler.setStackInSlot(realSlot, stack);
        }
    }

    @Override
    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
        var realSlot = slot - get(IoState.Output).getSlots();
        final var handler = get(IoState.Input);

        if (realSlot >= handler.getSlots() || realSlot < 0) return stack;
        return handler.insertItem(realSlot, stack, simulate);
    }

    @Override
    public ItemStack extractItem(int slot, int maxAmount, boolean simulate) {
        final var handler = get(IoState.Output);
        if (slot >= handler.getSlots() || slot < 0) return ItemStack.EMPTY;

        return handler.extractItem(
                slot,
                maxAmount,
                simulate
        );
    }

    @Override
    public int getSlotLimit(int slot) {
        final var handlerInput = get(IoState.Input);
        final var handlerOutput = get(IoState.Output);

        if (slot < 0) return 0;
        if (slot < handlerOutput.getSlots()) return handlerOutput.getSlotLimit(slot);
        if (slot < handlerInput.getSlots() + handlerOutput.getSlots()) return handlerInput.getSlotLimit(slot - handlerOutput.getSlots());
        return 0;
    }

    @Override
    public boolean isItemValid(int slot, ItemStack stack) {
        final var handlerInput = get(IoState.Input);
        final var handlerOutput = get(IoState.Output);

        if (slot < 0) return false;
        if (slot < handlerOutput.getSlots()) return handlerOutput.isItemValid(slot, stack);
        if (slot < handlerInput.getSlots() + handlerOutput.getSlots()) return handlerInput.isItemValid(slot - handlerOutput.getSlots(), stack);
        return false;
    }
}

File: src/main/kotlin/redcrafter07/processed/block/tile_entities/capabilities/MergedIoItemCapability.kt
package redcrafter07.processed.block.tile_entities.capabilities

import net.minecraft.world.item.ItemStack
import net.neoforged.neoforge.items.IItemHandler
import net.neoforged.neoforge.items.IItemHandlerModifiable
import net.neoforged.neoforge.items.wrapper.EmptyItemHandler
import redcrafter07.processed.block.machine_abstractions.IoState
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine

class MergedIoItemCapability(val handlers: ProcessedMachine.CapabilityHandlers) : IItemHandlerModifiable {
    private fun get(state: IoState): IItemHandler = handlers.getItemHandlerForState(state) ?: EmptyItemHandler.INSTANCE

    override fun getSlots(): Int {
        return get(IoState.Output).slots + get(IoState.Input).slots
    }

    override fun getStackInSlot(slot: Int): ItemStack {
        val handler = handlers.getItemHandlerForState(IoState.Output) ?: return ItemStack.EMPTY
        if (slot >= handler.slots || slot < 0) return ItemStack.EMPTY
        return handler.getStackInSlot(slot)
    }

    override fun setStackInSlot(slot: Int, stack: ItemStack) {
        val realSlot = slot - get(IoState.Output).slots

        val handler = get(IoState.Input)
        if (handler is IItemHandlerModifiable) {
            if (realSlot < 0 || realSlot >= handler.slots) return
            handler.setStackInSlot(realSlot, stack)
        }
    }

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        val realSlot = slot - get(IoState.Output).slots
        val handler = get(IoState.Input)

        if (realSlot >= handler.slots || realSlot < 0) return stack
        return handler.insertItem(realSlot, stack, simulate)
    }

    override fun extractItem(slot: Int, maxAmount: Int, simulate: Boolean): ItemStack {
        val handler = get(IoState.Output)
        if (slot >= handler.slots || slot < 0) return ItemStack.EMPTY

        return handler.extractItem(
            slot, maxAmount, simulate
        )
    }

    override fun getSlotLimit(slot: Int): Int {
        val handlerInput = get(IoState.Input)
        val handlerOutput = get(IoState.Output)

        if (slot < 0) return 0
        if (slot < handlerOutput.slots) return handlerOutput.getSlotLimit(slot)
        if (slot < handlerInput.slots + handlerOutput.slots) return handlerInput.getSlotLimit(slot - handlerOutput.slots)
        return 0
    }

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
        val handlerInput = get(IoState.Input)
        val handlerOutput = get(IoState.Output)

        if (slot < 0) return false
        if (slot < handlerOutput.slots) return handlerOutput.isItemValid(slot, stack)
        if (slot < handlerInput.slots + handlerOutput.slots) return handlerInput.isItemValid(
            slot - handlerOutput.slots, stack
        )
        return false
    }
}